orchagent/nexthopkey.h

```diff
@@ -15,9 +17,24 @@ struct NextHopKey
     string              alias;          // incoming interface alias
     uint32_t            vni;            // Encap VNI overlay nexthop
     MacAddress          mac_address;    // Overlay Nexthop MAC.
+    LabelStack          label_stack;    // MPLS label stack
```

Comment:
To support multiple nexthops with the same {prefix, label}, we also need to the mpls operation to the key.

Reply:
At the least actions will be "SWAP, PUSH,". You can follow the SAI_OUTSEG_TYPE to model the SAI label actions. POP action may be implicit in the label_stack being empty.

Reply:
@kperumalbfn @smaheshm No, this is not true for the model of MPLS supported by FRR and Linux (which is what this infra is meant for).  In Linux/FRR MPLS, there is always implicit POP of inseg and always implicit PUSH for non-empty label-stack in outseg NH.  SWAP in this scenario is really inseg POP followed by outseg PUSH.  There is no SWAP action on outseg NH.  There is no need for explicit  outseg action to be associated with NextHopKey.

Reply:
This is not going to scale well. If there are 100 different labels using one FRR, then this model would end up consuming 100 next hops.

Reply:
@smaheshm I don't understand this last comment.  How is this any different than scaling issues in IP-only routing with FRR?

Reply:
@smaheshm This is an inherent limitation of FRR/Linux.  There is no way in FRR/iproute2 CLI to specify MPLS route without POP on inseg label.  The only thing I can do is confine this limitation to fpmsyncd and expand APP_DB formatting to provide more flexibility for other/future routing stacks.   But at this point I would prefer to leave the expansion of MPLS to those who are integrating these other/future routing stacks and can provide the testing/support necessary.

Reply:
I'm guessing it's the iproute2 or libnl limitation, netlink protocol should allow it. Anyway we are using libnl so its ok to follow its logic. In my opinion specifying the action is a basic framework, if not's add now, later it adds too much overhead to add this feature, which can be avoided. There is no need to implement or test FRR. If you were to add the 'action' what major change are you anticipating. May be based on that we can decide.

Reply:
sai_outseg_type_t has been added to NextHopKey for this.